Profile View for issue #22 - Feature/22/profile screen

remivar opened the pull request:
created profile screen. The path Treetracker - UI - Views - Profile Screen

remivar committed:
added Profile Screen

remivar committed:
Added Profile Screen

remivar committed:
committing

remivar committed:
added frontend ui Profile Screen

remivar committed:
added front end ui Profile Screen

remivar committed:
committing profileviewcontroller, profileviewviewmodel

AlexCornforth committed:
Merge branch 'master' into feature/22/profile_screen

AlexCornforth committed:
Update AddTree.storyboard

Reverted AddTree storyboard changes

AlexCornforth committed:
Update AddTree.storyboard

Revert add tree storyboard changes

AlexCornforth committed:
Update AddTree.storyboard

Revert add tree storyboard changes

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewController.swift:
```diff
@@ -0,0 +1,64 @@
+    weak var rootcoordinator: RootCoordinator!
+    var homecoordinator: HomeCoordinator!
+    // MARK: - IBOutlets
+    @IBOutlet weak var profilepic: UIImageView!
```
Should we make this var private so it can't be access from outside the view?

AlexCornforth replied:
Also the variable name - we should try to avoid abreviations and also be consistent with the format of the other vriable names. Maybe something like `profileImageView` or `profilePictureImageView`would be better?

remivar replied:
> Should we make this var private so it can't be access from outside the view?

is that about rootcoordinator and homecoordinator

remivar replied:
I declared the properties rootcoordinator and homecoordinator as private. The name of IBoutlet for image is also changed

remivar replied:
I made the IBOutlet as private

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewController.swift:
```diff
@@ -0,0 +1,64 @@
+import UIKit
+
+class ProfileViewController: UIViewController {
+    var trees: [Tree] = []
```
I don't think we need to hold onto any tree objects in this view

remivar replied:
I removed the variable trees from ProfileViewController

remivar committed:
PR conflicts resolved

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewController.swift:
```diff
@@ -0,0 +1,64 @@
+            nameLabel.font = FontFamily.Montserrat.semiBold.font(size: 16.0)
+        }
+    }
+    @IBOutlet private var emailphoneLabel: UILabel! {
```
How about `usernameLabel` for the variable name?

remivar replied:
I changed the label name as usernameLabel

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewController.swift:
```diff
@@ -0,0 +1,64 @@
+            organizationLabel.font = FontFamily.Montserrat.semiBold.font(size: 16.0)
+        }
+    }
+    @IBAction func changeuserButton(_ sender: UIButton) {
```
Should we make this function private so it can't be accessed from outside the view?

AlexCornforth replied:
Also the function name could be a little more descriptive - `changeUserButtonPressed()` maybe? Keep an eye on the capitalisation.

As we're not using the sender value either maybe we should remove it?

remivar replied:
Made the changes in @IBAction func

AlexCornforth commented on TreeTracker/UI/Navigation/Home/HomeCoordinator.swift:
```diff
@@ -116,20 +117,27 @@ private extension HomeCoordinator {
+            viewModel.coordinatorDelegate = self
+            return viewModel
+        }()
         viewController.title = {
```
I don't think this logic should live in the coordinator. Maybe it best in the ViewModel what do you think?

remivar replied:
All Viewcontroller functions are written in HomeCoordinator. Eg: homeViewController, addTreeViewController, etc

AlexCornforth replied:
I'm referring to line 135 here - the title logic. We should move this to the ViewModel.

remivar replied:
I have moved the codes for title to ProfileViewModel

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+    weak var viewDelegate: ProfileViewModelViewDelegate?
+    weak var coordinatorDelegate: ProfileViewModelCoordinatorDelegate?
+    private let planter: Planter
+    private let treeMonitoringService: TreeMonitoringService
```
Interested in why you have chosen to use `TreeMonitoringService` in this ViewModel? It doesn't look like you are starting it anywhere so it wont do anything. Maybe consider removing along with the delegate methods if we don't need it.

remivar replied:
I removed TreeMonitoringService

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+        return planter.phoneNumber!
+    }
+    var planterOrganization: String {
+        return planter.organization!
```
The users organization is optional so its like this will often be nil so could cause a crash here. We should probably return an optional here.

remivar replied:
planter.organization is an optional. I don't understand what changes you requested here.

remivar replied:
gotcha. I defined the property planterOrganization as optional

AlexCornforth replied:
You are still force unwrapping `planter.organization` here. This will still crash if `planter.organization` is nil. You can jsut return the optional `planter.organization` value.

remivar replied:
I make the changes

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,92 @@
+            switch result {
+            case .success(let data):
+                guard let localPhotoPathImage = UIImage(data: data) else {
+                    fallthrough
```
Nice use of fallthrough here 👍

remivar replied:
Thank you

AlexCornforth reviewed (changes requested):
Thanks for working on this 👍 - I've left a few suggestions to work on.

Would also be good to revert all the changes unrelated to this change - storyboard files for example.

remivar committed:
Resolved the changes req in git

remivar committed:
Resolved all conversations in PR

remivar committed:
resolved addtree.storyboard

AlexCornforth commented on TreeTracker/UI/Views/Home/HomeViewModel.swift:
```diff
@@ -7,6 +7,7 @@
 //
 
 import UIKit
+import Foundation
```
We dont't need to import Foundation here as we import UIKit

remivar replied:
I removed Foundation from HomeViewModel. The codes for ProfileViewController's title has been added in the ProfileViewModel

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewController.swift:
```diff
@@ -0,0 +1,56 @@
+import UIKit
+
+class ProfileViewController: UIViewController {
+     weak var rootcoordinator: RootCoordinator!
```
We should remove the references to the coordinators here. The view layer shouldnt have knowledge of these - only the viewModel

remivar replied:
The references are removed

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,77 @@
+//  Copyright © 2021 Greenstand. All rights reserved.
+//
+
+import Foundation
```
No need to import Foundation here

remivar replied:
I have fixed those issues

AlexCornforth reviewed (changes requested):
Would be great if we could revert all the storyboard changes that are not related to this task.

remivar committed:
resolved the changes req in gitPR

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,79 @@
+        return "\(firstName) \(planter.lastName ?? "")"
+    }
+    var planterUsername: String {
+        guard planter.email == nil else {
```
if let would be more appropriate than a guard here as it will unwrap the email.

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,79 @@
+        guard planter.email == nil else {
+            return planter.email ?? ""
+        }
+        return planter.phoneNumber!
```
Lets remove the force unwrap here - could crash if phoneNumber is missing. I know its unlikely that both will be nil but just incase.

Think its fine to return an empty string in the unlikely event they are missing.

remivar replied:
I changed phoneNumber as non optional

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,79 @@
+        return "\(firstName) \(planter.lastName ?? "")"
+    }
+    var planterUsername: String {
+        guard planter.email == nil else {
```
`if let` would be more appropriate than a guard here as it will unwrap the email.

remivar replied:
I replaced guard statement with If-else

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,79 @@
+                guard let localPhotoPathImage = UIImage(data: data) else {
+                    fallthrough
+                }
+                let profiledetails = ProfileDetails(name: planterName, image: localPhotoPathImage, username: planterUsername, organization: planterOrganization!)
```
The force unwrap of organization will cause a crash here if organization is nil

remivar replied:
I used nil coalescing operator with  organization.

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,79 @@
+                let profiledetails = ProfileDetails(name: planterName, image: localPhotoPathImage, username: planterUsername, organization: planterOrganization!)
+                viewDelegate?.profileViewModel(self, didFetchDetails: profiledetails)
+            case .failure:
+                let profiledetails = ProfileDetails(name: planterName, image: Asset.Assets.person.image, username: planterUsername, organization: planterOrganization!)
```
The force unwrap of organization will cause a crash here if organization is nil

remivar replied:
used nil coalescing operator

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,79 @@
+        let organization: String
+    }
+    var title: String {
+        return "\(planter.firstName ?? "") \(planter.lastName ?? "")"
```
Maybe the title could be "Me" or "Profile" as before if both the first and last name are missing. Unlikely scenario but worth a graceful fallback.

remivar replied:
The SignUp button is inactive until the first name and last name are entered. However, I have made the changes requested here.

remivar committed:
resoled the changes req

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,83 @@
+}
+// MARK: - Profile
+extension ProfileViewModel {
+    var planterName: String {
```
These variables should prob ably be private as they are not used outside this class

remivar replied:
I made the three variables planterName, planterUsername, planterOrganization as private

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,83 @@
+import UIKit
+
+protocol ProfileViewModelCoordinatorDelegate: class {
+    func profileViewModel(_ profileViewModel: ProfileViewModel, changeUser planter: Planter)
```
Maybe update to: `func profileViewModel(_ profileViewModel: ProfileViewModel, didLogoutPlanter planter: Planter)` would be consistent with the rest of the app and swift partterns .

remivar replied:
changed the function name as didLogoutPlanter

remivar committed:
Update LaunchScreen.storyboard

remivar committed:
Update LaunchScreen.storyboard

remivar committed:
Update LaunchScreen.storyboard

remivar committed:
Update LaunchScreen.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Create AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update AddTree.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignIn.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update SignUp.storyboard

remivar committed:
Update Terms.storyboard

remivar committed:
Update Terms.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Home.storyboard

remivar committed:
Update Loading.storyboard

remivar committed:
Update Loading.storyboard

remivar committed:
Update Loading.storyboard

remivar committed:
Update Loading.storyboard

remivar committed:
Update Loading.storyboard

remivar committed:
Update Loading.storyboard

remivar committed:
resolved the changes req in PR

remivar committed:
resolved changes req in git pr

remivar committed:
committing in Xcode 12.5

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,83 @@
+        let organization: String
+    }
+    var title: String {
+        guard planter.firstName == nil, planter.lastName == nil else {
```
You could use `guard let` here as so:

```
guard let firstName = planter.firstName, let lastName = planter.lastName else {
    return L10n.Profile.FallbackTitle
}
return "\(firstName) \(lastName)"
```

remivar replied:
I did code changes

AlexCornforth commented on TreeTracker/UI/Views/Profile Screen/ProfileViewModel.swift:
```diff
@@ -0,0 +1,83 @@
+        guard planter.firstName == nil, planter.lastName == nil else {
+        return "\(planter.firstName ?? "") \(planter.lastName ?? "")"
+        }
+        return "My Profile"
```
We should add any user facing test to the Localizable strings file and access via the swiftgen `L10n` enum

remivar replied:
I added the Fallbacktitle in Localizable.strings and SwiftGen - Strings.swift

remivar committed:
add the fallback title of My Profile in SwiftGen and Localizable.strings, did the other code changes req in PR review.

remivar committed:
code changes req in pr

AlexCornforth committed:
Update HomeViewController.swift

Removed added whitespace

AlexCornforth committed:
Update Planter.swift

Undone live remoaval

AlexCornforth committed:
Update HomeCoordinator.swift

Removed unnecessary foundation import

AlexCornforth committed:
Update HomeCoordinator.swift

Removed whitespace

AlexCornforth committed:
Update HomeCoordinator.swift

Fixed spacing

AlexCornforth committed:
Update HomeCoordinator.swift

Removed unused tree monitoring service

AlexCornforth committed:
Update HomeCoordinator.swift

Nitpick: Fixed indentation for consistency

AlexCornforth committed:
Update SignIn.storyboard

Revert storyboard changes

AlexCornforth committed:
Update SignIn.storyboard

Revert storyboard changes

AlexCornforth committed:
Update Home.storyboard

Revert storyboard changes

AlexCornforth committed:
Update HomeViewModel.swift

Undone line removal

AlexCornforth committed:
Update ProfileViewController.swift

Removed unused AlertPresenting protocol conformance

AlexCornforth committed:
Update ProfileViewController.swift

Removed conformance to unused AlertPresenting protocol
Removed unused sender from button press function 
Moved button press to private extension
Fixed indentation

AlexCornforth committed:
Update ProfileViewModel.swift

Tidied up view controller:
- Removed force cast of planterUsername
- Made organization optional
- Create private extension for private variables
- Removed nil coalescing in title var

AlexCornforth committed:
Update ProfileViewModel.swift

Added function to create ProfileData rather than repeat code

AlexCornforth committed:
Update HomeCoordinator.swift

Nitpick: Spacing

AlexCornforth committed:
Update ProfileViewModel.swift

Added missing if

AlexCornforth committed:
Update ProfileViewModel.swift

Fixed optional issue in username

AlexCornforth commented:
@remivar I've made a bunch of changes directly on this PR. Would be good for you to take a look over them to understand what I've changed.
